kong/templates/nginx_kong.lua

```diff
@@ -284,7 +284,16 @@ server {
         default_type         '';
         set $kong_proxy_mode 'http';
 
-        rewrite_by_lua_block       {;}
+        rewrite_by_lua_block       {
```

Comment:
I think the upsteam tls info is only set before the first try. See [here](https://github.com/Kong/kong/blob/cb6dec46c07de556d285f602a5562762c45f968d/kong/runloop/balancer/init.lua#L306-L336).
As those upstream tls information is actually first set onto the the downstream `r`, then later when connecting to the upstream they are set onto the upstream connection `c` [here](https://github.com/Kong/lua-kong-nginx-module/blob/master/src/ngx_http_lua_kong_ssl.c#L51-L73).

Because `ngx.location.capture` emits a new request, what we've already set onto the old `r` gets lost. So basically what we do here, is setting those upstream tls information again onto the new `r`.

Reply:
No, balancer upstream TLS will reset in openresty balancer phase, Check code in here https://github.com/Kong/kong/blob/master/kong/init.lua#L1156

Reply:
The diagram below can illustrate it clearer

```
phase:    Kong.access.before              Kong.access.after               Kong.balancer(first try)    Kong.balancer(retry)
                 v                               v                                       |                 v
          balancer_prepare                balancer.execute(try_count=0)                  |            balancer.execute(try_count~=0)
                 v                               v                                       |                 |
          set service ssl config          set upstream ssl config                        v                 v
                           \                 /                                    init upstream ssl connection
                            \               /                                                   |
                             v             v                                                    v
                             ---------------                   set onto                     ------------
                             | downstream r|        ------------------------------>         |upstream c|
                             ---------------                                                ------------
```

Reply:
If we consider the `ngx.location.capture` case, then it will become

```
Kong.access.before            Kong.access.after             Kong.balancer(first try)    Kong.balancer(retry)
       v                             v                                     |                 v
balancer_prepare              balancer.execute(try_count=0)                |          balancer.execute(try_count~=0)
       v                             v                                     |                 |
set service ssl config        set upstream ssl config                      v                 v
                 \                /                                  init upstream ssl connection
                  \              /                                              |
                   v            v                                               v
                   ---------------                   set onto              ------------
                   | downstream r|        --------------------------->     |upstream c|
                   ---------------                                         ------------
ngx.location.capture     |
.........................+.............................................................................................
                         | ctx
                         | vars
                         v
                   ---------------                   set onto              ------------
                   |    sub r    |        --------------------------->     |upstream c|
                   ---------------                                         ------------
                         ^                                                      ^
                         |                                                      |
            -----------------------------                               ------------------
            |       Kong.rewrite        |                               |  Kong.balancer |
            |  set service ssl config   |                               |  same as above |
            |  set upstream ssl config  |                               ------------------
            -----------------------------
```

Reply:
Thanks! I has been understand it.